src/theory/strings/regexp_elim.cpp

```diff
@@ -0,0 +1,441 @@
+    // prev_end stores the current (symbolic) index in x that we are
+    // searching.
+    Node prev_end = d_zero;
+    unsigned gap_minsize_end = gap_minsize.back();
+    bool gap_exact_end = gap_exact.back();
+    std::vector<Node> non_greedy_find_vars;
+    for (unsigned i = 0, size = sep_children.size(); i < size; i++)
+    {
+      Node sc = sep_children[i];
+      if (gap_minsize[i] > 0)
+      {
+        // the gap to this child is at least gap_minsize[i]
+        prev_end =
+            nm->mkNode(PLUS, prev_end, nm->mkConst(Rational(gap_minsize[i])));
+      }
+      Node lensc = nm->mkNode(STRING_LENGTH, sc);
+      if (gap_exact[i])
+      {
+        // if the gap is exact, it is a substring constraint
+        Node curr = prev_end;
+        Node ss = nm->mkNode(STRING_SUBSTR, x, curr, lensc);
+        conj.push_back(ss.eqNode(sc));
+        prev_end = nm->mkNode(PLUS, curr, lensc);
+      }
+      else
+      {
+        // otherwise, we can use indexof to represent some next occurrence
+        if (gap_exact[i + 1] && i + 1 != size)
+        {
+          if (!options::regExpElimAgg())
+          {
+            success = false;
+            break;
+          }
+          // if the gap after this one is strict, we need a non-greedy find
+          // thus, we add a symbolic constant
+          Node k = nm->mkBoundVar(nm->integerType());
+          non_greedy_find_vars.push_back(k);
+          prev_end = nm->mkNode(PLUS, prev_end, k);
+        }
+        Node curr = nm->mkNode(STRING_STRIDOF, x, sc, prev_end);
+        Node idofFind = curr.eqNode(d_neg_one).negate();
+        conj.push_back(idofFind);
+        prev_end = nm->mkNode(PLUS, curr, lensc);
+      }
+      // if applicable, process the last gap
+      if (i == (size - 1))
+      {
+        if (gap_exact_end)
+        {
+          // substring relative to the end
+          Node curr = nm->mkNode(MINUS, lenx, lensc);
+          if (gap_minsize_end > 0)
+          {
+            curr =
+                nm->mkNode(MINUS, curr, nm->mkConst(Rational(gap_minsize_end)));
+          }
+          Node ss = nm->mkNode(STRING_SUBSTR, x, curr, lensc);
+          conj.push_back(ss.eqNode(sc));
+        }
+        else if (gap_minsize_end > 0)
+        {
+          Node fit = nm->mkNode(
+              LEQ,
+              nm->mkNode(
+                  PLUS, prev_end, nm->mkConst(Rational(gap_minsize_end))),
+              lenx);
+          conj.push_back(fit);
+        }
+      }
+    }
+    if (success)
+    {
+      // since sep_children is non-empty, conj is non-empty
+      Assert(!conj.empty());
+      Node res = conj.size() == 1 ? conj[0] : nm->mkNode(AND, conj);
+      // process the non-greedy find variables
+      if (!non_greedy_find_vars.empty())
+      {
+        std::vector<Node> children;
+        for (const Node& v : non_greedy_find_vars)
+        {
+          Node bound = nm->mkNode(
+              AND, nm->mkNode(LEQ, d_zero, v), nm->mkNode(LT, v, lenx));
```

Comment:
Shouldn't the upper bound be `lenx - gap_minsize_end` or something along the lines? We want to make sure that the last of the sep_children cannot overlap with the second to last of the sep_children.

Reply:
Abstractly, each of the non-greedy find variables k just means "I'm going to skip k indices in the string at this point". Based on the way the constraints are constructed, there is no remaining obligations on k at this point.  For example:

`x in re.++( re.*( re.allchar ), "A", re.allchar, "B", re.*( re.allchar ) )` gets converted to:

`indexof( x, "A", k ) != -1 ^ substr( x, indexof( x, "A", k )+2, 1 ) = "B"`, with non-greedy find variable k. At this point we just need to bound k, which can be bound to `0 <= k < len( x )`.

I'll add this as an example.

Reply:
The previous iteration of the code had an issue (I think) where the rewritten constraints could match more strings than the original constraints if there was an overlap between the last two `sep_children`. The new code resolves that.